app/src/main/java/ch/epfl/cs311/wanderwave/model/remote/BeaconConnection.kt

```diff
@@ -60,81 +60,51 @@ class BeaconConnection(
   }
 
   override fun getItem(itemId: String): Flow<Beacon> {
-    val dataFlow = MutableStateFlow<Beacon?>(null)
-
-    db.collection(collectionName)
-        .document(itemId)
-        .get()
-        .addOnSuccessListener { document ->
-          if (document != null && document.data != null) {
-            documentToItem(document)?.let { beacon ->
-              dataFlow.value = beacon
-
-              val tracksObject = document["tracks"]
-
-              var profileAndTrackRefs: List<Map<String, DocumentReference>>?
-
-              if (tracksObject is List<*> && tracksObject.all { it is Map<*, *> }) {
-                profileAndTrackRefs = tracksObject as? List<Map<String, DocumentReference>>
-
-                // Use a coroutine to perform asynchronous operations
-                coroutineScope.launch {
-                  val profileAndTracksDeferred =
-                      profileAndTrackRefs?.map { profileAndTrackRef ->
-                        async { fetchTrack(profileAndTrackRef) }
-                      }
-
-                  // Wait for all tracks to be fetched
-                  val profileAndTracks = profileAndTracksDeferred?.mapNotNull { it?.await() }
+    return getItem(itemId) { _, _ -> }
+  }
 
-                  // Update the beacon with the complete list of tracks
-                  val updatedBeacon = beacon.copy(profileAndTrack = profileAndTracks ?: emptyList())
-                  dataFlow.value = updatedBeacon
-                }
-              } else {
-                Log.e("Firestore", "tracks has Wrong Firebase Format")
-              }
+  override fun getItem(
+      itemId: String,
+      onSuccess: (DocumentSnapshot, MutableStateFlow<Beacon?>) -> Unit
+  ): Flow<Beacon> {
+    val onSuccessWrapper: (DocumentSnapshot, MutableStateFlow<Beacon?>) -> Unit =
+        { document, dataFlow ->
+          val beacon =
+              dataFlow.value
+                  ?: Beacon.from(document)
+                  ?: Beacon(

```

Comment:
Why are you doing this? Are you placing a beacon at the position (0,0) if we fail to retrieve the beacon?

Reply:
It's true, I'll return null in this case

Reply:
Solved